File: packages/rust/slop-ai/src/discovery/tools.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::tools::{affordances_to_tools, format_tree};

use super::service::DiscoveryService;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolContent {
    #[serde(rename = "type")]
    pub content_type: String,
    pub text: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolResult {
    pub content: Vec<ToolContent>,
    #[serde(skip_serializing_if = "is_false")]
    pub is_error: bool,
}

#[derive(Debug, Clone)]
pub struct DynamicToolEntry {
    pub name: String,
    pub description: String,
    pub input_schema: Value,
    pub provider_id: String,
    pub path: String,
    pub action: String,
}

#[derive(Debug, Clone)]
pub struct DynamicToolResolution {
    pub provider_id: String,
    pub path: String,
    pub action: String,
}

#[derive(Debug, Clone, Default)]
pub struct DynamicToolSet {
    pub tools: Vec<DynamicToolEntry>,
    resolve_map: HashMap<String, DynamicToolResolution>,
}

impl DynamicToolSet {
    pub fn resolve(&self, tool_name: &str) -> Option<&DynamicToolResolution> {
        self.resolve_map.get(tool_name)
    }
}

pub async fn create_dynamic_tools(service: &DiscoveryService) -> DynamicToolSet {
    let mut entries = Vec::new();
    let mut resolve_map = HashMap::new();

    for provider in service.get_providers().await {
        let Some(tree) = provider.consumer.tree(&provider.subscription_id).await else {
            continue;
        };

        let app_prefix = sanitize_prefix(&provider.id);
        let tool_set = affordances_to_tools(&tree, "");
        for tool in &tool_set.tools {
            let Some(resolution) = tool_set.resolve(&tool.function.name) else {
                continue;
            };

            let name = format!("{app_prefix}__{}", tool.function.name);
            entries.push(DynamicToolEntry {
                name: name.clone(),
                description: format!("[{}] {}", provider.name, tool.function.description),
                input_schema: tool.function.parameters.clone(),
                provider_id: provider.id.clone(),
                path: resolution.path.clone(),
                action: resolution.action.clone(),
            });
            resolve_map.insert(
                name,
                DynamicToolResolution {
                    provider_id: provider.id.clone(),
                    path: resolution.path.clone(),
                    action: resolution.action.clone(),
                },
            );
        }
    }

    DynamicToolSet {
        tools: entries,
        resolve_map,
    }
}

#[derive(Clone)]
pub struct ToolHandlers {
    service: DiscoveryService,
}

impl ToolHandlers {
    pub async fn list_apps(&self) -> ToolResult {
        let discovered = self.service.get_discovered().await;
        if discovered.is_empty() {
            return ToolResult {
                content: vec![ToolContent {
                    content_type: "text".to_string(),
                    text: "No applications found. Desktop and web apps that support external control will appear here automatically when they're running.".to_string(),
                }],
                is_error: false,
            };
        }

        let connected = self.service.get_providers().await;
        let mut connected_by_id = HashMap::new();
        for provider in connected {
            connected_by_id.insert(provider.id.clone(), provider);
        }

        let mut lines = Vec::new();
        for descriptor in discovered {
            let provider = connected_by_id.get(&descriptor.id);
            let tree = match provider {
                Some(provider) => provider.consumer.tree(&provider.subscription_id).await,
                None => None,
            };
            let action_count = tree
                .as_ref()
                .map(|tree| affordances_to_tools(tree, "").tools.len())
                .unwrap_or(0);
            let label = tree
                .as_ref()
                .and_then(|tree| tree.properties.as_ref())
                .and_then(|props| props.get("label"))
                .and_then(Value::as_str)
                .unwrap_or(&descriptor.name);
            let status = if provider.is_some() {
                format!("connected, {action_count} actions")
            } else {
                "available".to_string()
            };
            lines.push(format!(
                "- **{label}** (id: `{}`, {}) - {status}",
                descriptor.id, descriptor.transport.transport_type
            ));
        }

        ToolResult {
            content: vec![ToolContent {
                content_type: "text".to_string(),
                text: format!(
                    "Applications on this computer:\n{}\n\nUse connect_app with an app name or ID to connect and inspect it.",
                    lines.join("\n")
                ),
            }],
            is_error: false,
        }
    }

    pub async fn connect_app(&self, app: &str) -> ToolResult {
        match self.service.ensure_connected(app).await {
            Ok(Some(provider)) => {
                let Some(tree) = provider.consumer.tree(&provider.subscription_id).await else {
                    return ToolResult {
                        content: vec![ToolContent {
                            content_type: "text".to_string(),
                            text: format!("{} is connected but has no state yet.", provider.name),
                        }],
                        is_error: false,
                    };
                };

                let tool_set = affordances_to_tools(&tree, "");
                let actions = tool_set
                    .tools
                    .iter()
                    .map(|tool| {
                        let resolution = tool_set.resolve(&tool.function.name);
                        let action = resolution
                            .map(|resolution| resolution.action.as_str())
                            .unwrap_or(tool.function.name.as_str());
                        let path = resolution
                            .map(|resolution| resolution.path.as_str())
                            .unwrap_or("/");
                        format!("  - **{action}** on `{path}`: {}", tool.function.description)
                    })
                    .collect::<Vec<_>>()
                    .join("\n");

                ToolResult {
                    content: vec![ToolContent {
                        content_type: "text".to_string(),
                        text: format!(
                            "## {}\nID: `{}`\n\n### Current State\n```\n{}\n```\n\n### Available Actions ({})\n{}",
                            provider.name,
                            provider.id,
                            format_tree(&tree, 0),
                            tool_set.tools.len(),
                            actions,
                        ),
                    }],
                    is_error: false,
                }
            }
            Ok(None) => {
                let available = self
                    .service
                    .get_discovered()
                    .await
                    .into_iter()
                    .map(|descriptor| format!("{} ({})", descriptor.name, descriptor.id))
                    .collect::<Vec<_>>()
                    .join(", ");

                ToolResult {
                    content: vec![ToolContent {
                        content_type: "text".to_string(),
                        text: format!("App \"{app}\" not found. Available: {}", if available.is_empty() { "none" } else { &available }),
                    }],
                    is_error: true,
                }
            }
            Err(error) => ToolResult {
                content: vec![ToolContent {
                    content_type: "text".to_string(),
                    text: format!("Failed to connect to \"{app}\": {error}"),
                }],
                is_error: true,
            },
        }
    }

    pub async fn disconnect_app(&self, app: &str) -> ToolResult {
        if !self.service.disconnect(app).await {
            return ToolResult {
                content: vec![ToolContent {
                    content_type: "text".to_string(),
                    text: format!("App \"{app}\" is not connected. Use list_apps to see available apps."),
                }],
                is_error: true,
            };
        }

        ToolResult {
            content: vec![ToolContent {
                content_type: "text".to_string(),
                text: format!("Disconnected from \"{app}\". Its tools have been removed."),
            }],
            is_error: false,
        }
    }
}

pub fn create_tool_handlers(service: DiscoveryService) -> ToolHandlers {
    ToolHandlers { service }
}

fn sanitize_prefix(value: &str) -> String {
    value
        .chars()
        .map(|c| if c.is_ascii_alphanumeric() { c } else { '_' })
        .collect::<String>()
        .trim_matches('_')
        .to_string()
}

fn is_false(value: &bool) -> bool {
    !value
}

File: packages/go/slop-ai/discovery/tools.go
package discovery

import (
	"context"
	"fmt"
	"strings"

	slop "github.com/devteapot/slop/packages/go/slop-ai"
)

// ToolContent is a text content block for host integrations.
type ToolContent struct {
	Type string `json:"type"`
	Text string `json:"text"`
}

// ToolResult is a host-agnostic tool result payload.
type ToolResult struct {
	Content []ToolContent `json:"content"`
	IsError bool          `json:"isError,omitempty"`
}

// DynamicToolEntry describes a single affordance-backed tool.
type DynamicToolEntry struct {
	Name        string
	Description string
	InputSchema any
	ProviderID  string
	Path        string
	Action      string
}

// DynamicToolResolution resolves a dynamic tool name back to invoke coordinates.
type DynamicToolResolution struct {
	ProviderID string
	Path       string
	Action     string
}

// DynamicToolSet contains all affordance-backed tools across connected providers.
type DynamicToolSet struct {
	Tools      []DynamicToolEntry
	resolveMap map[string]DynamicToolResolution
}

// Resolve maps a dynamic tool name back to invoke coordinates.
func (d DynamicToolSet) Resolve(toolName string) (DynamicToolResolution, bool) {
	resolution, ok := d.resolveMap[toolName]
	return resolution, ok
}

// CreateDynamicTools builds namespaced tool definitions from all connected providers.
func CreateDynamicTools(service *Service) DynamicToolSet {
	entries := []DynamicToolEntry{}
	resolveMap := map[string]DynamicToolResolution{}

	for _, provider := range service.GetProviders() {
		tree := provider.Consumer.Tree(provider.SubscriptionID)
		if tree == nil {
			continue
		}

		prefix := sanitizePrefix(provider.ID)
		toolSet := slop.AffordancesToTools(*tree, "")
		for _, tool := range toolSet.Tools {
			resolution, ok := toolSet.Resolve(tool.Function.Name)
			if !ok {
				continue
			}

			name := prefix + "__" + tool.Function.Name
			entries = append(entries, DynamicToolEntry{
				Name:        name,
				Description: fmt.Sprintf("[%s] %s", provider.Name, tool.Function.Description),
				InputSchema: tool.Function.Parameters,
				ProviderID:  provider.ID,
				Path:        resolution.Path,
				Action:      resolution.Action,
			})
			resolveMap[name] = DynamicToolResolution{
				ProviderID: provider.ID,
				Path:       resolution.Path,
				Action:     resolution.Action,
			}
		}
	}

	return DynamicToolSet{Tools: entries, resolveMap: resolveMap}
}

// ToolHandlers exposes the core discovery tool handlers.
type ToolHandlers struct {
	service *Service
}

// CreateToolHandlers builds the host-agnostic core tool handlers.
func CreateToolHandlers(service *Service) ToolHandlers {
	return ToolHandlers{service: service}
}

// ListApps lists all discovered providers and connection state.
func (h ToolHandlers) ListApps() ToolResult {
	discovered := h.service.GetDiscovered()
	if len(discovered) == 0 {
		return ToolResult{
			Content: []ToolContent{{
				Type: "text",
				Text: "No applications found. Desktop and web apps that support external control will appear here automatically when they're running.",
			}},
		}
	}

	connected := h.service.GetProviders()
	connectedByID := map[string]*ConnectedProvider{}
	for _, provider := range connected {
		connectedByID[provider.ID] = provider
	}

	lines := make([]string, 0, len(discovered))
	for _, desc := range discovered {
		provider := connectedByID[desc.ID]
		tree := (*slop.WireNode)(nil)
		actionCount := 0
		if provider != nil {
			tree = provider.Consumer.Tree(provider.SubscriptionID)
			if tree != nil {
				actionCount = len(slop.AffordancesToTools(*tree, "").Tools)
			}
		}

		label := desc.Name
		if tree != nil {
			if value, ok := tree.Properties["label"].(string); ok && value != "" {
				label = value
			}
		}

		status := "available"
		if provider != nil {
			status = fmt.Sprintf("connected, %d actions", actionCount)
		}

		lines = append(lines, fmt.Sprintf("- **%s** (id: `%s`, %s) - %s", label, desc.ID, desc.Transport.Type, status))
	}

	return ToolResult{
		Content: []ToolContent{{
			Type: "text",
			Text: fmt.Sprintf("Applications on this computer:\n%s\n\nUse connect_app with an app name or ID to connect and inspect it.", strings.Join(lines, "\n")),
		}},
	}
}

// ConnectApp connects to a provider and returns its current state snapshot and actions.
func (h ToolHandlers) ConnectApp(ctx context.Context, app string) ToolResult {
	provider, err := h.service.EnsureConnected(ctx, app)
	if err != nil {
		return ToolResult{Content: []ToolContent{{Type: "text", Text: fmt.Sprintf("Failed to connect to %q: %v", app, err)}}, IsError: true}
	}
	if provider == nil {
		available := make([]string, 0, len(h.service.GetDiscovered()))
		for _, desc := range h.service.GetDiscovered() {
			available = append(available, fmt.Sprintf("%s (%s)", desc.Name, desc.ID))
		}
		return ToolResult{Content: []ToolContent{{Type: "text", Text: fmt.Sprintf("App %q not found. Available: %s", app, strings.Join(available, ", "))}}, IsError: true}
	}

	tree := provider.Consumer.Tree(provider.SubscriptionID)
	if tree == nil {
		return ToolResult{Content: []ToolContent{{Type: "text", Text: fmt.Sprintf("%s is connected but has no state yet.", provider.Name)}}}
	}

	toolSet := slop.AffordancesToTools(*tree, "")
	actions := make([]string, 0, len(toolSet.Tools))
	for _, tool := range toolSet.Tools {
		resolution, ok := toolSet.Resolve(tool.Function.Name)
		action := tool.Function.Name
		pathInfo := ""
		if ok {
			action = resolution.Action
			pathInfo = fmt.Sprintf(" on `%s`", resolution.Path)
		}
		actions = append(actions, fmt.Sprintf("  - **%s**%s: %s", action, pathInfo, tool.Function.Description))
	}

	return ToolResult{
		Content: []ToolContent{{
			Type: "text",
			Text: fmt.Sprintf("## %s\nID: `%s`\n\n### Current State\n```\n%s\n```\n\n### Available Actions (%d)\n%s", provider.Name, provider.ID, slop.FormatTree(*tree, 0), len(toolSet.Tools), strings.Join(actions, "\n")),
		}},
	}
}

// DisconnectApp disconnects an active provider connection.
func (h ToolHandlers) DisconnectApp(app string) ToolResult {
	if !h.service.Disconnect(app) {
		return ToolResult{Content: []ToolContent{{Type: "text", Text: fmt.Sprintf("App %q is not connected. Use list_apps to see available apps.", app)}}, IsError: true}
	}
	return ToolResult{Content: []ToolContent{{Type: "text", Text: fmt.Sprintf("Disconnected from %q. Its tools have been removed.", app)}}}
}

func sanitizePrefix(value string) string {
	var b strings.Builder
	lastUnderscore := false
	for _, r := range value {
		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') {
			b.WriteRune(r)
			lastUnderscore = false
			continue
		}
		if !lastUnderscore && b.Len() > 0 {
			b.WriteByte('_')
			lastUnderscore = true
		}
	}
	return strings.Trim(b.String(), "_")
}
